### DISCUSSION TITLE
Memory leak on RaspberryPi 4B

### DISCUSSION THREAD

**Discussion by dfybc** [aiortc/aiortc#742] [Q&A] [answered]
when i run the example  webcam on raspberry 4B(4G ram),  i found the memory occupancy increases until memory is consumed.

i have noticed that someone had submit such a issue, but now is invalid,  

I hope I can help with this problem。

**Comment by chschnell** [comment]
I can reproduce this reliably with camera.py on a Raspberry Pi Zero 2W.

My setup:
- aioice and aiortc are cloned from github (not the PIP packages), built and installed with "setup.py install --user".
- USB camera: Microsoft® LifeCam HD-3000 (connected through an active USB hub)
- Camera resoultion: 320x240
- OS: current Raspberry Pi OS

Open two SSH sessions to the Pi, start `camera.py` in one and `htop -p <PROCESS_ID_OF_CAMERA_PY>` in the other.

If you connect just with a single browser session, camera.py is stable for at least 24 hours (didn't test beyond that).

If you open multiple browser tabs and connect to camera.py in parallel you will trigger this bug, it get's a bit undeterministic here but 2 browser tabs can be enough, else try with 3. It also helps a lot if you disconnect and reconnect one browser tab while the other(s) keep playing.

Once the bug hits (which is basically immediately once you play around with multiple browser sessions), I see leakage in htop of about 2M/second, since the Pi Zero 2W has only half a gig it takes around 2 minutes until the linux kernel kills the python interpreter runnnig camera.py. If I disconnect all sessions before linux kills the process, the leakage stops but memory leaked up until this point is also not returned to the os.

I ran the python interpreter with tracemalloc hoping to get some info on where the leakage could roughly occur, but the output did not seem to include the leaked memory, I guess it happens in one of the C libraries invisible to the interpreter. But then I'm not sure if I operate tracemalloc correctly, maybe somebody else can confirm this.

I don't know if it is linux-, raspberrypi- or even armhf-specific. However I cannot reproduce this bug on my intel Windows machine. It might also be the case that this is triggered due to the Pi not having enough computational power to keep up (though I did try with 1 fps, didn't help), my Windows machine is definitely fast enough (there I run it with 640x480 with no problems whatsoever).

**Comment by neilyoung** [comment]
It's a while ago, but did this lead to some solution? I'm having this same issue: I'm just subscribing to an offered video, doing nothing with the track. In the beginning I forgot to assign the received track to the peerConnection, this drove the memory usage over the top.

But even if I just accept it - the memory of the PI is slowly consumed. No  `gc.collect` helps. Only termination of the app, this frees the memory again.

**Reply by jlaine** [reply, marked as answer]
@neilyoung what do you mean "doing nothing with the track"? If you don't consume the incoming media frames, for sure they are just going to build up and run you out of memory. That's why the examples use `MediaBlackhole` which reads the media and discards it.

**Reply by neilyoung** [reply]
@jlaine I think this issue is solved. I simply forgot to attache a MediaBlackHole or something to the track. Works now
